src/Microsoft.DotNet.Darc/src/DarcLib/Helpers/LocalPath.cs

```diff
@@ -0,0 +1,121 @@
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+namespace Microsoft.DotNet.DarcLib.Helpers;
+
+/// <summary>
+/// Smart file path wrapper that allows easy combining and makes sure separators conform to the desired way.
+/// They support concatenation where the left-most (root) path decides the shape of the outcome.
+/// (e.g. "C:\foo" / "src/something" returns "C:\foo\src\something")
+/// 
+/// The concatenation works thanks to overloading of the / operator:
+///     var combinedPath = rootPath / "src" / "MyProject.csproj";
+/// </summary>
+public abstract class LocalPath
+{
+    private readonly char _separator;
+
+    public string Path { get; }
+
+    public int Length => Path.Length;
+
+    public LocalPath(string rootPath, char separator) : this(rootPath, separator, true)
+    {
+    }
+
+    protected LocalPath(string path, char separator, bool normalizePath)
+    {
+        Path = normalizePath ? NormalizePath(path) : path;
+        _separator = separator;
+    }
+
+    public override string ToString() => Path;
+
+    public static LocalPath operator /(LocalPath left, LocalPath right) => left.CreateMergedPath(left.Combine(left.Path, left.NormalizePath(right.Path)));
+
+    public static LocalPath operator /(LocalPath left, string right) => left.CreateMergedPath(left.Combine(left.Path, left.NormalizePath(right)));
+
+    public static LocalPath operator /(string left, LocalPath right) => right.CreateMergedPath(right.Combine(right.NormalizePath(left), right.Path));
+
+    public static implicit operator string(LocalPath p) => p.Path;
+
+    protected abstract LocalPath CreateMergedPath(string path);
+
+    protected abstract string NormalizePath(string s);
+
+    private string Combine(string left, string right)
+    {
+        var slashCount = (left.EndsWith(_separator) ? 1 : 0) + (right.StartsWith(_separator) ? 1 : 0);
+
+        return slashCount switch
+        {
+            0 => left + _separator + right,
+            1 => left + right,
+            2 => left + right[1..],
+            _ => throw new System.NotImplementedException(),
+        };
+    }
+
+    public override bool Equals(object? obj) => Path.Equals((obj as LocalPath)?.Path ?? obj as string);
+
+    public override int GetHashCode() => Path.GetHashCode();
+}
+
+/// <summary>
+/// Smart path that defaults to using whatever is the current directory separator.
+/// </summary>
+public class NativePath : LocalPath
+{
+    public NativePath(string rootPath) : this(rootPath, true)
+    {
+    }
+
+    private NativePath(string rootPath, bool normalize) : base(rootPath, System.IO.Path.DirectorySeparatorChar, normalize)
+    {
+    }
+
+    protected override LocalPath CreateMergedPath(string path) => new NativePath(path, false);
+
+    protected override string NormalizePath(string s) =>
+        System.IO.Path.DirectorySeparatorChar == '/'
+            ? s.Replace('\\', '/')
```

Comment:
Since these separator chars are used all over the place, it might be good to abstract them to constants.

Reply:
I'd do that for magic strings but not sure it will help anything for slash characters?

Reply:
i.e. I think `/` is much easier than `UnixDirectorySeparator`?